### PR TITLE
#42 - fix decompose(X) and add doctest

### PR DESCRIPTION
**Pull request by mforets**

### CODE REVIEW

**Comment by schillic** [comment]
Can we add that `decompose(X, eps==Inf)` falls back to this faster function?

File: src/Approximations/decompositions.jl [hunk 1]
```diff
@@ -63,30 +63,31 @@ function decompose(X::LazySet)::CartesianProductArray
     n = LazySets.dim(X)
     b = div(n, 2)
 
-    DIR_EAST_b = repeat(DIR_EAST, outer=b)
-    DIR_NORTH_b = repeat(DIR_NORTH, outer=b)
-    DIR_WEST_b = repeat(DIR_WEST, outer=b)
-    DIR_SOUTH_b = repeat(DIR_SOUTH, outer=b)
+    DIR_EAST_bi = Vector{Float64}(n)
```
Comment (mforets) [inline src/Approximations/decompositions.jl]: these should be `sparsevec([], Float64[], n)`, because below they are assigned sparse vectors

Comment (schillic) [reply src/Approximations/decompositions.jl]: Then change it? :)

Comment (mforets) [reply src/Approximations/decompositions.jl]: hmm actually we may just use a single vector for east/north, and another one for west/south, by circular shifting it.

File: src/Approximations/decompositions.jl [hunk 2]
```diff
@@ -12,6 +12,8 @@ Return an approximation of the given 2D set as a box-shaped polygon.
 A polygon in constraint representation.
 """
 function overapproximate(X::LazySet)::HPolygon
+    @assert dim(X) == 2
```
Comment (schillic) [inline src/Approximations/decompositions.jl]: 👍

File: src/Approximations/decompositions.jl (hunk 1 above)
Comment (mforets) [reply src/Approximations/decompositions.jl]: nah, i tried this but it has roughly the same performance:

```julia
    DIR_EAST_bi = sparsevec([1], [1.], n)
    DIR_NORTH_bi = similar(DIR_EAST_bi)
    DIR_WEST_bi = sparsevec([1], [-1.], n)
    DIR_SOUTH_bi = similar(DIR_WEST_bi)
    result = Vector{HPolygon}(b)

    @inbounds for bi in 1:b
        pe_bi = dot(DIR_EAST, view(σ(DIR_EAST_bi, X), 2*bi-1:2*bi))
        circshift!(DIR_NORTH_bi, DIR_EAST_bi, 1)
        pn_bi = dot(DIR_NORTH, view(σ(DIR_NORTH_bi, X), 2*bi-1:2*bi))
        copy!(DIR_EAST_bi, DIR_NORTH_bi)

        pw_bi = dot(DIR_WEST, view(σ(DIR_WEST_bi, X), 2*bi-1:2*bi))
        circshift!(DIR_SOUTH_bi, DIR_WEST_bi, 1)
        ps_bi = dot(DIR_SOUTH, view(σ(DIR_SOUTH_bi, X), 2*bi-1:2*bi))
        copy!(DIR_WEST_bi, DIR_SOUTH_bi)

        result[bi] = HPolygon([LinearConstraint(DIR_EAST, pe_bi),
                               LinearConstraint(DIR_NORTH, pn_bi),
                               LinearConstraint(DIR_WEST, pw_bi),
                               LinearConstraint(DIR_SOUTH, ps_bi)])
    end
```

**Comment by mforets** [comment]
then it will be ready to merge when the CI finishes

**Pull request merged**